### DISCUSSION TITLE
Panda classes overriden by non-layered CSS

### DISCUSSION THREAD

**Discussion by 0ubbe** [chakra-ui/panda#1728] [Ideas]
### Description

First of all, thanks for the fantastic work with this library, it is a great initiative 🙌🏽 

I'm trying to migrate our Chakra application to **Panda** + **Ark UI**. We want to use RSC, and having to wait for Chakra to have real support for that eventually is an overkill for us 💀 

Adding **Panda** to our NextJS app was painless following the docs. We also re-used Chakra's theme into Panda (with some minor adjustments).

The problem we're encountering, is that **Chakra UI** uses [Emotion](https://emotion.sh/), which does a typical top-level CSS reset:

<img width="1071" alt="Screenshot 2023-11-29 at 17 27 07" src="https://github.com/chakra-ui/panda/assets/5938217/e13a2841-10b0-4f7d-aff2-6c85ca84b5b7">

That ends up applying some styles to HTML native elements:

<img width="342" alt="Screenshot 2023-11-29 at 17 24 11" src="https://github.com/chakra-ui/panda/assets/5938217/0d1400c0-c3eb-4ea5-beca-abfd13b58f17">

When we try to style a button in our app with Panda, we run into a specificity issue 😢 

<img width="337" alt="Screenshot 2023-11-29 at 17 24 16" src="https://github.com/chakra-ui/panda/assets/5938217/9e4044ce-f468-496b-88d0-b436f3aa2a9c">

All Panda's utilities (and styles) are declared in layers; any top-level CSS easily overrides those.

While it is excellent Panda[ is using layers](https://developer.mozilla.org/en-US/docs/Web/CSS/@layer), it feels those can be easily, unintentionally overridden by external CSS from a framework or library...

It would be beneficial if we could make opt-in out of layers in Panda:
```ts
{
  useLayers: false
}
```
So that we can migrate existing apps with non-layered CSS out of our control easily. Our app is big, and if we aim to migrate out of Chakra to Panda we need to do it step-by-step.

### Link to Reproduction

https://chakra-ui.com/getting-started/nextjs-guide

### Steps to reproduce

1. Install [Chakra UI with NextJS](https://chakra-ui.com/getting-started/nextjs-guide)
2. Install Panda
3. Try to style a button via Panda in the NextJS app changing its `background-color`
4. See the styles not being applied due to specificity

### JS Framework

NextJS

### Panda CSS Version

0.19.0

### Browser

Any

### Operating System

- [X] macOS
- [X] Windows
- [X] Linux

**Comment by astahmer** [comment]
hey, there are a few solutions to choose from:
- you can scope the Chakra-UI css reset with `<ChakraProvider resetScope=".some-scope">`
- you can remove Panda CSS @layers with a postcss plugin https://www.npmjs.com/package/@csstools/postcss-cascade-layers
- IF your Panda styles are contained to a specific part of your app, not mixed with Chakra styles, (e.g a widget), you can make use of the shadow DOM to completely isolate the styling, with react it's as easy as wrapping your component with `react-shadow`

I don't think removing layers would "just work" due to how the CSS is currently generated (it's not deterministically generated in the ideal order = the same as the one provided by the @layers), this is addressed in https://github.com/chakra-ui/panda/pull/1544 but even then I'm not sure it would still behave 100% correctly, that's why I recommend using the postcss plugin linked above if you really need to remove layers, since it emulated the layers specificities by adding things in the selector

all that being said, I don't feel like this is really a Panda issue but rather an outside source spreading global styles that should instead be scoped somehow, hopefully here this is easily possible with Chakra `resetScope`

**Reply by 0ubbe** [reply]
Thanks for the thorough answer 🙏🏽 🙇🏽 . I'll give a go to the proposed solutions and let you know how it went 🙌🏽

**Comment by JuhG** [comment]
Running into this, I think a better solution is to put the styles on the global space, outside of any layers. We write our main CSS with Panda, these are not "utilities", these are the app level code we want to ship. I don't see any reason to lower specificity on them. Any global CSS coming from a random package being able to overwrite them seems wrong. But also nesting it under body introduces another set of issues :)

A better solution is just removing the layer. I would love a first party solution, something like:
```ts
{
  layers: {
    utilities: false,
  },
}
```
But asking on discord core team didn't seem like the idea.

Here's out solution, based on recommendation from @astahmer.

Create a helper that moves all styles out of `utilities` layer:
```ts
const removeUtilitiesLayer = (css: string) => {
  const root = postcss.parse(css);

  root.walkAtRules("layer", (atRule) => {
    if (atRule.params === "utilities") {
      const parent = atRule.parent;

      const nodes = atRule.nodes;
      if (nodes) {
        nodes.forEach((node) => {
          parent.insertBefore(atRule, node.clone());
        });
      }

      atRule.remove();
    }
  });

  return root.toString();
};
```

Use it in your panda config:
```ts
{
  hooks: {
    "cssgen:done": ({ artifact, content }) => {
      if (artifact === "styles.css") {
        content = removeUtilitiesLayer(content);
      }
      return content;
    },
  },
}
```

**Comment by whoisYeshua** [comment]
Bump
